### PR TITLE
Add ability to construct the equivalent FIAT element

### PR DESCRIPTION
**Pull request by miklos1**
- Adds `.fiat_equivalent` property that gives the equivalent FIAT element for any particular FInAT element.
- Mostly implemented as `cached_property`, so as to simulate the caching feature of the FIAT interface in TSFC.

### CODE REVIEW

File: finat/enriched.py [hunk 1]
```diff
@@ -71,12 +73,11 @@ def fiat_equivalent(self):
...
-            from FIAT.mixed import MixedElement  # on-demand loading
-            return MixedElement([e.element.fiat_equivalent
-                                 for e in self.elements], ref_el=self.cell)
+            return FIAT.MixedElement([e.element.fiat_equivalent
+                                      for e in self.elements], ref_el=self.cell)
         else:
-            from FIAT.enriched import EnrichedElement  # on-demand loading
-            return EnrichedElement(*[e.fiat_equivalent for e in self.elements])
+            return FIAT.EnrichedElement(*[e.fiat_equivalent
+                                          for e in self.elements])
```
[the remark this answers is not in the document]
Comment (miklos1) [reply finat/enriched.py]: Done.

File: finat/finiteelementbase.py [hunk 2]
```diff
@@ -99,6 +99,14 @@ def index_shape(self):
     def value_shape(self):
         '''A tuple indicating the shape of the element.'''
 
+    @property
+    def fiat_equivalent(self):
+        '''The FIAT element equivalent to this FInAT element.'''
+        raise NotImplementedError(str.format(
+            "Cannot make equivalent FIAT element for {classname}",
+            classname=type(self).__name__
+        ))
```
[the remark this answers is not in the document]
Comment (miklos1) [reply finat/finiteelementbase.py]: I don't really like f-strings, but whatever...

File: finat/nodal_enriched.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+from finat.fiat_elements import FiatElement
+
+
+class NodalEnrichedElement(FiatElement):
+    """An enriched element with a nodal basis."""
+
+    def __init__(self, elements):
+        fiat_elements = [elem.fiat_equivalent for elem in elements]
+        nodal_enriched = FIAT.NodalEnrichedElement(*fiat_elements)
```
[the remark this answers is not in the document]
Comment (miklos1) [reply finat/nodal_enriched.py]: Done.

File: finat/fiat_elements.py [hunk 4]
```diff
@@ -267,6 +272,12 @@ def value_shape(self):
         return ()
 
 
+class Bernstein(ScalarFiatElement):
+    # TODO: Replace this with a smarter implementation
+    def __init__(self, cell, degree):
+        super(Bernstein, self).__init__(FIAT.Bernstein(cell, degree))
```
[the remark this answers is not in the document]
Comment (miklos1) [reply finat/fiat_elements.py]: I did it this way for stylistic consistency with the surround code, but I agree `super().__init__(...)` is better: with that one does not forget changing the class name here when the class name above is changed (e.g. due to copy-paste, etc.).

**Comment by dham** [comment]
This looks like a straight addition of features. Is there any reason not to merge this right away?

**Comment by miklos1** [comment]
> This looks like a straight addition of features. Is there any reason not to merge this right away?

Currently waiting for FEniCS/fiat#57 because of discontinuous Raviart-Thomas.  If no discussion happens there in a few days, I will proceed with Lawrence's proposal and then merge.

File: finat/discontinuous.py [hunk 5]
```diff
@@ -45,6 +45,20 @@ def index_shape(self):
...
+    def fiat_equivalent(self):
+        from FIAT.discontinuous import DiscontinuousElement
+        from FIAT.discontinuous_raviart_thomas import DiscontinuousRaviartThomas
+        from FIAT.raviart_thomas import RaviartThomas
+
+        fiat_element = self.element.fiat_equivalent
+        if isinstance(fiat_element, RaviartThomas):
+            ref_el = fiat_element.get_reference_element()
+            deg = fiat_element.degree()
+            return DiscontinuousRaviartThomas(ref_el, deg)
```
[the remark this answers is not in the document]
Comment (miklos1) [reply finat/discontinuous.py:58]: > OTOH, we probably want the latter. So I vote for removing that test (and probably culling DRT from FIAT).

I am doing this.

**Comment by miklos1** [comment]
> Currently waiting for [FEniCS/fiat#57](https://github.com/FEniCS/fiat/issues/57) because of discontinuous Raviart-Thomas. 

In absence of discussion on DRT element, I suggest merging this PR as it is **now.**  (We can still change things later if a better resolution is achieved.)

**Comment by miklos1** [comment]
Apparently, I cannot merge without an approve tick.  (And ticks are dismissed when any changes are pushed.)

**Review by wence- — APPROVED** [review]

**Comment by wence-** [comment]
> Apparently, I cannot merge without an approve tick. (And ticks are dismissed when any changes are pushed.)

The latter is reasonably deliberate, I can probably fix the former. But this is merged, thanks!

**Comment by miklos1** [comment]
> But this is merged, thanks!

Thanks.  I sync now the tsfc and firedrake branches with upstream.  Those two should be merged together (at the same time), but I think I can do that myself.

**Pull request merged**